Fix for RNN/LSTM/GRU modules to work with stateless.functional_call()

jbschlosser opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #91111

Fixes #90500

The change here checks for parameter changes at the beginning of each `forward()` call; if the parameters are found to be different tensors than last time, `self._flat_weights` is re-initialized with the new values. Thus, swapping parameter values using `stateless.functional_call()` will re-initialize `self._flat_weights` during the `forward()` call, and the provided parameters will be used for module computation as expected.

NB: There are still some changes needed for symbolic shapes to work with `nn.GRU` (will address in a follow-up PR).

cc @albanD @mruberry @walterddr @kshitij12345 @saketh-are

jbschlosser committed:
Fix for RNN/LSTM/GRU modules to work with stateless.functional_call()

[ghstack-poisoned]

jbschlosser committed:
Update on "Fix for RNN/LSTM/GRU modules to work with stateless.functional_call()"


Fixes #90500

The change here checks for parameter changes at the beginning of each `forward()` call; if the parameters are found to be different tensors than last time, `self._flat_weights` is re-initialized with the new values. Thus, swapping parameter values using `stateless.functional_call()` will re-initialize `self._flat_weights` during the `forward()` call, and the provided parameters will be used for module computation as expected.

cc albanD mruberry walterddr kshitij12345 saketh-are

[ghstack-poisoned]

jbschlosser committed:
Update on "Fix for RNN/LSTM/GRU modules to work with stateless.functional_call()"


Fixes #90500

The change here checks for parameter changes at the beginning of each `forward()` call; if the parameters are found to be different tensors than last time, `self._flat_weights` is re-initialized with the new values. Thus, swapping parameter values using `stateless.functional_call()` will re-initialize `self._flat_weights` during the `forward()` call, and the provided parameters will be used for module computation as expected.

NB: There are still some changes needed for symbolic shapes to work with `nn.GRU` (will address in a follow-up PR).

cc albanD mruberry walterddr kshitij12345 saketh-are

[ghstack-poisoned]

ezyang reviewed (approved):
ahhh, this is very good, thank you!

albanD commented on torch/nn/modules/rnn.py:
```diff
@@ -258,6 +274,12 @@ def extra_repr(self) -> str:
 
+    def __getstate__(self):
+        # Don't serialize the weight references.
+        state = self.__dict__.copy()
```
should this be `state = super().__getstate__()` ?

jbschlosser replied:
I tried that first, but apparently `super().__getstate__()` didn't exist until Python 3.11 for some reason.

AFAICT from the python docs, the default behavior when a `__getstate__()` impl is not present is equivalent to returning the object's `state_dict`.

albanD commented on torch/nn/modules/rnn.py:
```diff
@@ -268,33 +290,32 @@ def __setstate__(self, d):
+                        else:
+                            self._all_weights += [weights[:2]]
+                            self._flat_weights_names.extend(weights[:2])
+        self._init_flat_weights()
```
The old code for this function was not calling `self.flatten_parameters()`?

jbschlosser replied:
You're correct about this. It seemed incomplete to me since the other two places this is done do include a call to `flatten_parameters()`, but I'm happy to revert it back just in case.

albanD commented on torch/nn/modules/rnn.py:
```diff
@@ -123,11 +125,16 @@ def __init__(self, mode: str, input_size: int, hidden_size: int,
         self.reset_parameters()
 
+    def _init_flat_weights(self):
+        self._flat_weights = [(lambda wn: getattr(self, wn) if hasattr(self, wn) else None)(wn) for wn in self._flat_weights_names]
```
Not yours but:
```
self._flat_weights = [getattr(self, wn, None) for wn in self._flat_weights_names]
```

jbschlosser replied:
Haha good call, super weird to use a lambda this way

jbschlosser replied:
So mypy does not like the use of `getattr(self, wn, None)`. i can still get rid of the lambda though

albanD reviewed (approved):
Looks good!
Only small nits

jbschlosser committed:
Update on "Fix for RNN/LSTM/GRU modules to work with stateless.functional_call()"


Fixes #90500

The change here checks for parameter changes at the beginning of each `forward()` call; if the parameters are found to be different tensors than last time, `self._flat_weights` is re-initialized with the new values. Thus, swapping parameter values using `stateless.functional_call()` will re-initialize `self._flat_weights` during the `forward()` call, and the provided parameters will be used for module computation as expected.

NB: There are still some changes needed for symbolic shapes to work with `nn.GRU` (will address in a follow-up PR).

cc albanD mruberry walterddr kshitij12345 saketh-are

[ghstack-poisoned]

jbschlosser committed:
Update on "Fix for RNN/LSTM/GRU modules to work with stateless.functional_call()"


Fixes #90500

The change here checks for parameter changes at the beginning of each `forward()` call; if the parameters are found to be different tensors than last time, `self._flat_weights` is re-initialized with the new values. Thus, swapping parameter values using `stateless.functional_call()` will re-initialize `self._flat_weights` during the `forward()` call, and the provided parameters will be used for module computation as expected.

NB: There are still some changes needed for symbolic shapes to work with `nn.GRU` (will address in a follow-up PR).

cc albanD mruberry walterddr kshitij12345 saketh-are

[ghstack-poisoned]

jbschlosser commented:
@pytorchbot merge

jbschlosser commented:
@pytorchbot merge -f "Unrelated failures"
